modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/sorted/PageMemorySortedIndexStorage.java

```diff
@@ -247,28 +233,20 @@ public void close() {
 
             @Override
             public boolean hasNext() {
-                if (!closeBusyLock.enterBusy()) {
-                    throwStorageClosedException();
-                }
+                return busy(() -> {
+                    throwExceptionIfStorageInProgressOfRebalance(state.get(), PageMemorySortedIndexStorage.this::createStorageInfo);
 
-                try {
                     return cursor.hasNext();
-                } finally {
-                    closeBusyLock.leaveBusy();
-                }
+                });
```

Comment:
Is this code really easier to read then the previous one?

Reply:
Also, but why not use a generic template for this class?

Reply:
Why force it?